packages/vm/lib/buildBlock.ts

```diff
@@ -0,0 +1,241 @@
+    for (const [i, txResult] of this.transactionResults.entries()) {
+      const tx = this.transactions[i]
+      gasUsed.iadd(txResult.gasUsed)
+      const { encodedReceipt } = await generateTxReceipt.bind(this.vm)(tx, txResult, gasUsed)
+      await receiptTrie.put(encode(i), encodedReceipt)
+    }
+    return receiptTrie.root
+  }
+
+  /**
+   * Adds the block miner reward to the coinbase account.
+   */
+  private async rewardMiner() {
+    const minerReward = new BN(this.vm._common.param('pow', 'minerReward'))
+    const reward = calculateMinerReward(minerReward, 0)
+    const coinbase = this.headerData.coinbase
+      ? new Address(toBuffer(this.headerData.coinbase))
+      : Address.zero()
+    await rewardAccount(this.vm.stateManager, coinbase, reward)
+  }
+
+  /**
+   * Run and add a transaction to the block being built.
+   * Please note that this modifies the state of the VM.
+   * Throws if the transaction's gasLimit is greater than
+   * the remaining gas in the block.
+   */
+  async addTransaction(tx: TypedTransaction) {
+    this.checkStatus()
+
+    if (!this.checkpointed) {
+      await this.vm.stateManager.checkpoint()
+      this.checkpointed = true
+    }
+
+    // According to the Yellow Paper, a transaction's gas limit
+    // cannot be greater than the remaining gas in the block
+    const blockGasLimit = new BN(toBuffer(this.headerData.gasLimit))
+    const blockGasRemaining = blockGasLimit.sub(this.gasUsed())
+    if (tx.gasLimit.gt(blockGasRemaining)) {
+      throw new Error('tx has a higher gas limit than the remaining gas in the block')
+    }
+
+    const blockData = { header: this.headerData, transactions: this.transactions }
+    const block = Block.fromBlockData(blockData, this.blockOpts)
+
+    const result = await this.vm.runTx({ tx, block })
+
+    this.transactions.push(tx)
+    this.transactionResults.push(result)
+
+    return result
+  }
+
+  /**
+   * Reverts the checkpoint on the StateManager to reset the state from any transactions that have been run.
+   */
+  async revert() {
+    this.checkStatus()
+    if (this.checkpointed) {
+      await this.vm.stateManager.revert()
+      this.reverted = true
+    }
+  }
+
+  /**
+   * This method returns the finalized block.
+   * It also:
+   *  - Assigns the reward for miner (PoW)
+   *  - Commits the checkpoint on the StateManager
+   *  - Sets the tip of the VM's blockchain to this block
+   * For PoW, optionally seals the block with params `nonce` and `mixHash`,
+   * which is validated along with the block number and difficulty by ethash.
+   * For PoA, please pass `blockOption.cliqueSigner` into the buildBlock constructor,
+   * as the signer will be awarded the txs amount spent on gas as they are added.
+   */
+  async build(sealOpts?: SealBlockOpts) {
+    this.checkStatus()
+    const blockOpts = this.blockOpts
+    const consensusType = this.vm._common.consensusType()
+
+    if (consensusType === 'pow') {
+      await this.rewardMiner()
+    }
+
```

Comment:
If we use clique, is the miner rewarded?

Reply:
there are no block rewards in clique poa. the signer will get the txs amount spent on gas (applied in `runTx`)